util/src/main/java/tc/oc/pgm/util/nms/NMSHacks.java

```diff
@@ -42,847 +32,277 @@
+  NMSHacksPlatform INSTANCE = chooseNMSHacks();
 
-  static void sendPacket(Player bukkitPlayer, Object packet) {
-    if (bukkitPlayer.isOnline()) {
-      EntityPlayer nmsPlayer = ((CraftPlayer) bukkitPlayer).getHandle();
-      nmsPlayer.playerConnection.sendPacket((Packet) packet);
-    }
-  }
+  static NMSHacksPlatform chooseNMSHacks() {
+    NMSHacksPlatform choice;
+    Logger logger = Logger.getLogger("NMSHacks");
 
-  static void sendPacketToViewers(Entity entity, Object packet) {
-    sendPacketToViewers(entity, packet, false);
-  }
-
-  static void sendPacketToViewers(Entity entity, Object packet, boolean excludeSpectators) {
-    EntityTrackerEntry entry = getTrackerEntry(entity);
-    for (EntityPlayer viewer : ((Set<EntityPlayer>) entry.trackedPlayers)) {
-      if (excludeSpectators) {
-        Entity spectatorTarget = viewer.getBukkitEntity().getSpectatorTarget();
-        if (spectatorTarget != null && spectatorTarget.getUniqueId().equals(entity.getUniqueId()))
-          continue;
-      }
-      viewer.playerConnection.sendPacket((Packet) packet);
-    }
-  }
-
-  /** Immediately send the given entity's metadata to all viewers in range */
-  static void sendEntityMetadataToViewers(Entity entity, boolean complete) {
-    sendPacketToViewers(entity, entityMetadataPacket(entity.getEntityId(), entity, complete));
-  }
-
-  Constructor<PacketPlayOutPlayerInfo.PlayerInfoData> playerInfoDataConstructor =
-      getPlayerInfoDataConstructor();
-
-  static Constructor<PacketPlayOutPlayerInfo.PlayerInfoData> getPlayerInfoDataConstructor() {
     try {
-      Constructor<PacketPlayOutPlayerInfo.PlayerInfoData> constructor =
-          PacketPlayOutPlayerInfo.PlayerInfoData.class.getConstructor(
-              PacketPlayOutPlayerInfo.class,
-              GameProfile.class,
-              int.class,
-              WorldSettings.EnumGamemode.class,
-              IChatBaseComponent.class);
-
-      constructor.setAccessible(true);
-      return constructor;
-    } catch (NoSuchMethodException e) {
-      throw new RuntimeException(e);
-    }
-  }
-
-  static PacketPlayOutPlayerInfo.PlayerInfoData playerListPacketData(
-      PacketPlayOutPlayerInfo packet,
-      UUID uuid,
-      String name,
-      GameMode gamemode,
-      int ping,
-      @Nullable Skin skin,
-      @Nullable String renderedDisplayName) {
-    GameProfile profile = new GameProfile(uuid, name);
-    if (skin != null) {
-      for (Map.Entry<String, Collection<Property>> entry :
-          Skins.toProperties(skin).asMap().entrySet()) {
-        profile.getProperties().putAll(entry.getKey(), entry.getValue());
+      if (BukkitUtils.isSportPaper()) {
+        choice = new NMSHacksSportPaper();
+        logger.info("Using NMSHacksSportPaper");
+      } else {
+        choice = new NMSHacks1_8();
+        logger.info("Using NMSHacks1_8");
       }
+    } catch (Throwable t1) {
+      choice = new NMSHacksNoOp();
+      logger.warning("Using Fallback NMSHacksNoOp!");
     }
```

Comment:
```suggestion
      if (BukkitUtils.isSportPaper()) {
        choice = new NMSHacksSportPaper();
      } else {
        choice = new NMSHacks1_8();
      }
      logger.info("Using " + choice.getClass().getSimpleName());
    } catch (Throwable t1) {
      choice = new NMSHacksNoOp();
      logger.log(Level.WARNING, t1, () -> "Using Fallback NMSHacksNoOp!");
    }
```

Can't remember if the log entry includes the throwable in the terminal by default, but nice to include if it is :)

Reply:
Would the error actually be useful?
It's just going to print class not found, which is expected if you aren't using 1.8 and try to use 1.8 NMS code.